pkg/jobs/remove_link.go

```diff
@@ -0,0 +1,110 @@
+)
+
+type RemoveLinkJob struct {
+	accountId uint64
+	linkId    uint64
+	userId    uint64
+	log       *logrus.Entry
+	db        *pg.DB
+}
+
+func (r *RemoveLinkJob) Run(ctx context.Context) error {
+	span := sentry.StartSpan(ctx, "Job", sentry.TransactionName("Remove Link"))
+	defer span.Finish()
+
+	log := r.log
+
+	return r.db.RunInTransaction(span.Context(), func(txn *pg.Tx) error {
+		repo := repository.NewRepositoryFromSession(r.userId, r.accountId, txn)
+
+		link, err := repo.GetLink(span.Context(), r.linkId)
+		if err != nil {
+			crumbs.Warn(span.Context(), "failed to retrieve link to be removed, this job will not be retried", "weirdness", nil)
+			log.WithError(err).Error("failed to retrieve link that to be removed, this job will not be retried")
+			return nil
+		}
+
+		// Delete the link directly, I don't want to include something like this on the repository interface as it is not
+		// something that I want just anything to be able to do. Deleting a link has the potential to move a ton of data
+		// and should be done in the background. This should do a cascade if I setup my foreign keys correctly but tests
+		// should be written to verify that those cascades are _always_ happening properly.
+		result, err := txn.ModelContext(span.Context(), link).
+			WherePK().
+			Delete(&link)
+		if err != nil {
+			log.WithError(err).Error("failed to delete link")
+			return err
+		}
+
+		log.WithField("rowsAffects", result.RowsAffected()).Info("successfully removed link")
+
+		return nil
+	})
+}
+
+func (j *jobManagerBase) newRemoveLinkJob(job *work.Job) (*RemoveLinkJob, error) {
+	log := j.getLogForJob(job)
+
+	// We need to know what link we are actually deleting. This is required.
+	linkId := uint64(job.ArgInt64("linkId"))
+	if linkId == 0 {
+		log.Error("link Id is 0, the link Id must be specified for removal")
+		return nil, errors.New("must specify link Id for removal")
+	}
+
+	// Deleting links is never an automatic process. This is always initiated by a user, so we want to include the user
+	// who initiatied this action in our logs so we have some record of _who_ initiated the delete.
+	userId := uint64(job.ArgInt64("userId"))
+	if userId == 0 {
+		log.Error("user Id is 0, the user Id must be specified for removal")
+		return nil, errors.New("must specify user Id for removal")
+	}
+
+	// Validate that we also have an accountId.
+	accountId, err := j.getAccountId(job)
+	if err != nil {
+		log.WithError(err).Error("could not run job, no account Id")
+		return nil, err
+	}
+
+	log = log.WithFields(logrus.Fields{
+		"accountId": accountId,
+		"linkId":    linkId,
+		"userId":    userId,
+	})
+
+	runner := &RemoveLinkJob{
+		accountId: accountId,
+		linkId:    linkId,
+		userId:    userId,
+		log:       log,
+		db:        j.db,
+	}
+
+	return runner, nil
+}
+
+func (j *jobManagerBase) removeLink(input *work.Job) error {
+	job, err := j.newRemoveLinkJob(input)
+	if err != nil {
+		return err
+	}
+
+	return job.Run(context.Background())
```

Comment:
I'm not sure what the implications of this would be. But maybe it would be best to pass a context with a deadline?